sys/kern/kern_zcond.c

```diff
@@ -0,0 +1,382 @@
+MALLOC_DEFINE(M_ZCOND, "zcond", "malloc for the zcond subsystem");
+
+static void
+zcond_load_patch_points(linker_file_t lf)
+{
+	struct patch_point *begin, *end;
+	struct patch_point *ins_p;
+	struct zcond *owning_zcond;
+
+	if (linker_file_lookup_set(lf, __XSTRING(ZCOND_LINKER_SET), &begin,
+		&end, NULL) == 0) {
+		for (ins_p = begin; ins_p < end; ins_p++) {
+			owning_zcond = ins_p->zcond;
+
+			if (owning_zcond->patch_points.slh_first == NULL) {
+				SLIST_INIT(&owning_zcond->patch_points);
+			}
+
+			SLIST_INSERT_HEAD(&owning_zcond->patch_points, ins_p,
+			    next);
+		}
+	}
+}
+
+static void
+zcond_kld_load(void *arg __unused, struct linker_file *lf)
+{
+	zcond_load_patch_points(lf);
+}
+
+static int
+zcond_load_patch_points_cb(linker_file_t lf, void *arg __unused)
+{
+	zcond_load_patch_points(lf);
+	return (0);
+}
+
+/*
+ * Collect patch_points from the __zcond_table ELF section into a list.
+ * Prepare a CPU local copy of the kernel_pmap, used to safely patch
+ * an instruction.
+ */
+static vm_offset_t mirror_addr;
+static void
+zcond_init(const void *unused)
+{
+	EVENTHANDLER_REGISTER(kld_load, zcond_kld_load, NULL,
+	    EVENTHANDLER_PRI_ANY);
+	linker_file_foreach(zcond_load_patch_points_cb, NULL);
+	mirror_addr = zcond_get_patch_va();
+}
+SYSINIT(zcond, SI_SUB_ZCOND, SI_ORDER_SECOND, zcond_init, NULL);
+
+struct zcond_patch_arg {
+	int patching_cpu;
+	struct zcond *cond;
+	struct zcond_md_ctxt *md_ctxt;
+	bool enable;
+};
+
+/*
+ * Patch all patch_points belonging to cond.
+ */
+static void
+zcond_patch(struct zcond *cond, bool enable)
+{
+	struct patch_point *p;
+	vm_page_t patch_page;
+	uint8_t insn[ZCOND_MAX_INSN_SIZE];
+	size_t insn_size;
+
+	SLIST_FOREACH(p, &cond->patch_points, next) {
+		zcond_get_patch_insn(p, insn, &insn_size);
+
+		patch_page = PHYS_TO_VM_PAGE(vtophys(p->patch_addr));
+		pmap_qenter_zcond(patch_page);
+
+		zcond_before_patch();
+		memcpy((void *)(mirror_addr + (p->patch_addr & PAGE_MASK)),
+		    &insn[0], insn_size);
+		zcond_after_patch();
+	}
+}
+
+static void
+rendezvous_setup(void *arg)
+{
+	struct zcond_patch_arg *data;
+
+	data = (struct zcond_patch_arg *)arg;
+
+	if (data->patching_cpu == curcpu) {
+		zcond_before_rendezvous(data->md_ctxt);
+	}
+}
+
+static void
+rendezvous_action(void *arg)
+{
+	struct zcond_patch_arg *data;
+
+	data = (struct zcond_patch_arg *)arg;
+
+	if (data->patching_cpu == curcpu) {
+		zcond_patch(data->cond, data->enable);
+	}
+}
+
+static void
+rendezvous_teardown(void *arg)
+{
+	struct zcond_patch_arg *data;
+
+	data = (struct zcond_patch_arg *)arg;
+
+	if (data->patching_cpu == curcpu) {
+		zcond_after_rendezvous(data->md_ctxt);
+	}
+}
+
+void
+__zcond_toggle(struct zcond *cond, bool enable)
+{
+	struct zcond_md_ctxt ctxt;
+
+	if (enable && refcount_acquire(&cond->refcnt) > 1) {
+		return;
+	} else if (!enable && !refcount_release_if_not_last(&cond->refcnt)) {
+		return;
+	}
+
+	if (!enable && refcount_load(&cond->refcnt) > 1) {
+		return;
+	}
```

Comment:
```suggestion
	if (enable && refcount_acquire(&cond->refcnt) > 0) {
		return;
	} else if (!enable && !refcount_release_if_not_last(&cond->refcnt)) {
		return;
	}

```

Ovo bi trebalo pokriti sve slučajeve, nekako više smisla ima da se `refcnt` incijalizira na 0.
Onaj originalni snippet koji sam predložio je trebao radit tako, ali sam slučajno dodao +1 na sve checkove.